### DISCUSSION TITLE
cellPlugin examples

### DISCUSSION THREAD

**Discussion by zehawki** [react-page/react-page#913] [Show and tell]
### Add a MaterialUI button:

```
import Mouse from '@material-ui/icons/MouseTwoTone'

// Do standard materialUI stuff here - import Button, makeStyles, set useStyles

export const buttonPlugin = {
  id: 'button-plugin',
  title: 'Button',
  description: 'Add a button',
  version: 1,
  icon: <Mouse />,

  Renderer: ({ data }) => {
    const classes = useStyles();

    // Transform any link into one that is relative and resolves to this site only
    const baseURL = window.location.protocol + "//" + window.location.hostname
    const url = new URL(data.relativeURL, baseURL);
    const relativeURL = url.pathname

    // Use whatever deepCopy foo you like. I use a simple one like this
    // export function deepCopy(array) {
    //   return JSON.parse(JSON.stringify(array))
    // }

    var localData = deepCopy(data)
    if (localData.color)
      localData.color = localData.color.toLowerCase()
    if (localData.size)
      localData.size = localData.size.toLowerCase()
    if (localData.design)
      localData.design = localData.design.toLowerCase()

    return (
      <div className={classes.root}>
        <Link to={relativeURL}>
          <Button color={localData.color} size={localData.size} variant={localData.design} >
            {data.buttonText}
          </Button>
        </Link>
      </div>
    )
  },

  controls: {
    type: 'autoform',
    schema: {
      properties: {
        buttonText: {
          type: 'string',
          default: 'Launch network',
        },
        relativeURL: {
          type: 'string',
          default: '/network',
        },
        color: {
          type: 'string',
          default: 'primary',
          enum: ['Primary', 'Secondary'],
        },
        size: {
          type: 'string',
          default: 'medium',
          enum: ['Small', 'Medium', 'Large'],
        },
        design: {
          type: 'string',
          default: 'contained',
          enum: ['Bare', 'Outlined', 'Contained'],
        },
      },
      required: ['buttonText', 'relativeURL', 'color', 'size', 'design'],
    },
  },
};
```

**Comment by macrozone** [comment]
I would recommend to use a custom slate plugin for Links and similar, so that the LinkButton can also be place inline into text.

see https://react-page.github.io/docs/#/slate?id=slate-plugins-with-custom-data

`var localData = deepCopy(data)`

1.) don't use deepCopy, just use object spread, but in this case i would just pass it directly to the Button
2.) I see why you did lowercase the label, you can actually specify  transform function on the underlying `uniforms`-library: [check the uniforms playground](https://uniforms.tools/playground/#?N4IgDgTgpgzlAuIBcICCATd0YwAQAoApAZQHkA5YgYwAsoBbAQwEoQAacCAezBmVEYBXeFxiMAblGQAzRgBs4HISLGSAIlDmMAnsgCMABgMd0ASzEAjOVHQz5ikFoubk8CIKgcwWqlBpc5dCgIOwVPEGhGdFIAOzldJFkwjhhaBkZkEHx8ZlwAXgA-XGAAHRjcXCouGJh4XEYAK3F83BioAHdcVCb8YHq5OQBRCG4IGCRcNw82XEE4DVlBOXhxyfdPXABrKG12rgh0VYBtEpBBGNNpffoYU4BdXABfZgBuMoqqmrrUuiYW0vKFUmpng1gmAHIMFhYDBwWx3kD4NowFAIVwLA0oFR4HCERVIDxgvBTLAJgCgUCrlwyUiURNTrUIKYYgBzU4zKAxQT0CYnECMdm4U4WQWnKj3GbnS7XVbANyMGpXCA83DNQqqgB0IgAqmAURAAMKMOA5R6PPFPeGAirQACOglM0HQvPBjKgCDhuHBAC9TGBPa74Ix4FBwXcrRVHm8YgjpOdsaZqpVIiGAGryUzoYP7fA_dK5clAz61VUZrMiCAtRriDVVehgUzWXNpJivMoW6DwQQQcr4ehcIJyXLqwsU8Rl7MQPsDzRtmPWoGXAjjuSZyca4KjPAAMm3pdX5f2G5G-xgGusrPgNALFopne75T6QSDjdWK7XFePW6e0YpkYtUYIua84fNUJZ5kw6YHpOLRUCmUBQR-OYQSw0YIvePatB0uAkBQ1C_IwABCTLoCyUDNgRMwoYhh4QHOzw5OwIAwP47QAJJxMyUDDKMoQOMa2gxFQsTEIIFj0CCfHhAJQmxDRwZSIk9hQI8HBXgwikgBYXBcCs8pgAALCAjxAA
)

in your case you would specify on `enum` what values are actually possible to store and for displaying it, you can use a transform function and pass it through e.g. a translation library, uppercasing or whatever:

```
size: {
          type: 'string',
          default: 'medium',
          enum: ['small', 'medium', 'large'],
          uniforms: {
              transform: v => uppercaseFirst(v) // or whatever you like
          }
        },

```

**Reply by zehawki** [reply]
> I would recommend to use a custom slate plugin for Links and similar, so that the LinkButton can also be place inline into text.

I thought about that, but wanted to have full control over placement of major buttons (which are CTAs)

> you can actually specify transform function on the underlying

Ah, thank you! I thought as much that uniforms must surely be having a way, but didnt have the time to explore yet.
